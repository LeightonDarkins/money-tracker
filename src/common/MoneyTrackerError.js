class MoneyTrackerError {
  constructor (status = 'NETWORK ERROR', reason, message) {
    this.id = Date.now()
    this.status = status
    this.reason = reason
    this.message = message
  }

  getTitle () {
    if (this.status === 'NETWORK ERROR') return this.status

    return `${this.status} ${this.reason}`
  }

  getMessage () {
    if (this.status === 'NETWORK ERROR') return 'There was a problem connecting to the server. Check your internet connection and try again.'

    return this.message
  }

  getId () {
    return this.id
  }
}

module.exports = MoneyTrackerError
