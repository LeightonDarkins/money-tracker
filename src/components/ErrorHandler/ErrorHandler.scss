@import '../../variables.scss';

#ErrorHandler {
  z-index: 999;
  position: absolute;
  width: 100vw;
  max-width: 1280px;
  margin: 0;
}
