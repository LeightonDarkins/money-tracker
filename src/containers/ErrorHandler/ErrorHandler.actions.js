export const types = {
  apiError: 'API_ERROR',
  closeError: 'CLOSE_ERROR'
}

export const apiError = error => ({ type: types.apiError, error })

export const closeError = error => ({ type: types.closeError, error })
