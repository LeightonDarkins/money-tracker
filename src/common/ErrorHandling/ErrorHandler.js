import MoneyTrackerError from './MoneyTrackerError'

export const ErrorHandler = (error) => {
  if (error.response) {
    const response = error.response

    let message = response.data ? response.data.message : 'An unknown error occured.'

    return new MoneyTrackerError(
      response.status,
      response.statusText,
      message)
  } else {
    return new MoneyTrackerError()
  }
}
