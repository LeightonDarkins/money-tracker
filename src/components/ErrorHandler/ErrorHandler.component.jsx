import React, { Component } from 'react'
import PropTypes from 'prop-types'
import ErrorComponent from '../Error/Error.component.jsx'

import './ErrorHandler.scss'

class ErrorHandler extends Component {
  render () {
    return (<div id='ErrorHandler'>
      { this.props.errors.map(errorItem => {
        return <ErrorComponent
          key={errorItem.getId()}
          title={errorItem.getTitle()}
          message={errorItem.getMessage()}
          onClick={() => this.props.onErrorClick(errorItem.getId())}
        />
      }) }
    </div>)
  }
}

ErrorHandler.propTypes = {
  errors: PropTypes.array.isRequired
}

export default ErrorHandler
