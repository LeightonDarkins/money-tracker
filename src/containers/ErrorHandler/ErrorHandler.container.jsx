import { connect } from 'react-redux'

import ErrorHandler from '../../components/ErrorHandler/ErrorHandler.component.jsx'
import { closeError } from './ErrorHandler.actions.js'

const mapStateToProps = state => {
  return {
    errors: state.ErrorHandler.errors
  }
}

const mapDispatchToProps = dispatch => {
  return {
    onErrorClick: (key) => {
      dispatch(closeError(key))
    }
  }
}

const ConnectedErrorHandler = connect(
  mapStateToProps,
  mapDispatchToProps)(ErrorHandler)

export default ConnectedErrorHandler
