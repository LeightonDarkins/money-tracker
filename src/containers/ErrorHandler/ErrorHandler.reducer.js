import { types } from './ErrorHandler.actions'

const initialState = { errors: [] }

const accounts = (state = initialState, action) => {
  switch (action.type) {
    case types.apiError:
      if (state.errors.length > 0) {
        return Object.assign({}, state, { errors: [...state.errors, action.error] })
      }

      return Object.assign({}, state, { errors: [action.error] })
    case types.closeError:
      let filteredErrors = state.errors.filter(err => err.id !== action.error)

      return Object.assign({}, state, { errors: filteredErrors })

    default:
      return state
  }
}

export default accounts
