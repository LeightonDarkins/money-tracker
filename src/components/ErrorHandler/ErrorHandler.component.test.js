/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import ErrorHandler from './ErrorHandler.component'
import th from '../../common/TestHelpers'
import MoneyTrackerError from '../../common/MoneyTrackerError'

describe('ErrorHandler', () => {
  const errors = [
    new MoneyTrackerError()
  ]

  describe('Snapshots', () => {
    it('should render an empty ErrorHandler with no errors', () => {
      const errorHandler = renderer.create(<ErrorHandler
        errors={[]}
        onErrorClick={th.empty}
        />).toJSON()

      expect(errorHandler).toMatchSnapshot()
    })

    it('should render a populated ErrorHandler with errors', () => {
      const errorHandler = renderer.create(<ErrorHandler
        errors={errors}
        onErrorClick={th.empty}
        />).toJSON()

      expect(errorHandler).toMatchSnapshot()
    })
  })

  describe('Behaviour', () => {
    let mockOnErrorClick

    beforeEach(() => {
      mockOnErrorClick = jest.fn()
    })

    afterEach(() => {
      mockOnErrorClick.mockReset()
    })

    it('calls onErrorClick when an Error is clicked', () => {
      let wrapper = shallow(<ErrorHandler
        errors={errors}
        onErrorClick={mockOnErrorClick} />)

      wrapper.find('ErrorComponent').simulate('click')

      expect(mockOnErrorClick).toBeCalled()
    })
  })
})
