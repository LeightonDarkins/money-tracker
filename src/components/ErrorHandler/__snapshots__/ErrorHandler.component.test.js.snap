exports[`ErrorHandler Snapshots should render a populated ErrorHandler with errors 1`] = `
<div
  id="ErrorHandler">
  <div
    className="ErrorComponent"
    onClick={[Function]}>
    <h3
      className="heading">
      NETWORK ERROR
    </h3>
    <p
      className="body">
      There was a problem connecting to the server. Check your internet connection and try again.
    </p>
    <i
      aria-hidden="true"
      className="fa fa-times close" />
  </div>
</div>
`;

exports[`ErrorHandler Snapshots should render an empty ErrorHandler with no errors 1`] = `
<div
  id="ErrorHandler" />
`;
